### PR TITLE
feat: add `StatementStatistics` for slow query logging implementation

### PR DESCRIPTION
**Pull request by zyy17**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

Resolve https://github.com/GreptimeTeam/greptimedb/issues/4450

The PR introduces the new struct `StatementStatistics` for implementing statistics of query statements. For now, it's used for slow query logging.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

**Comment by sunng87** [comment]
Is there any chance we can put this at instance level, say `SqlQueryHandler` and `PrometheusHandler`? so we don't need to configure it for each protocol.

**Comment by zyy17** [comment]
> Is there any chance we can put this at instance level, say `SqlQueryHandler` and `PrometheusHandler`? so we don't need to configure it for each protocol.

Good idea. Let me try to do that.

File: config/config.md [hunk 1]
```diff
@@ -162,6 +162,9 @@
 | `logging.log_format` | String | `text` | The log format. Can be `text`/`json`. |
 | `logging.tracing_sample_ratio` | -- | -- | The percentage of tracing will be sampled and exported.<br/>Valid range `[0, 1]`, 1 means all traces are sampled, 0 means all traces are not sampled, the default value is 1.<br/>ratio > 1 are treated as 1. Fractions < 0 are treated as 0 |
 | `logging.tracing_sample_ratio.default_ratio` | Float | `1.0` | -- |
+| `logging.slow_query` | -- | -- | The slow query log options. |
+| `logging.slow_query.enable` | Bool | `false` | Whether to enable slow query log. |
+| `logging.slow_query.threshold` | String | Unset | The threshold of slow query. |
```
Comment (killme2008) [inline config/config.md:167]: Except the threshold, maybe we need a sample rate too. Just like mongodb 

https://www.mongodb.com/zh-cn/docs/cloud-manager/performance-advisor/slow-query-threshold/

File: src/common/telemetry/src/logging.rs [hunk 2]
```diff
@@ -235,6 +252,42 @@ pub fn init_global_logging(
             None
         };
 
+        let slow_query_logging_layer = if !opts.dir.is_empty() && opts.slow_query.enable {
+            let rolling_appender =
+                RollingFileAppender::new(Rotation::HOURLY, &opts.dir, "greptimedb-slow");
```
Comment (killme2008) [inline src/common/telemetry/src/logging.rs:257]: Looks better
```suggestion
                RollingFileAppender::new(Rotation::HOURLY, &opts.dir, "greptimedb-slow-queries");
```

File: src/servers/src/http/handler.rs [hunk 3]
```diff
@@ -138,7 +138,19 @@ pub async fn sql(
     if let Some(limit) = query_params.limit {
         resp = resp.with_limit(limit);
     }
-    resp.with_execution_time(start.elapsed().as_millis() as u64)
+
+    let cost = start.elapsed();
+    if let Some(threshold) = state.slow_query_threshold {
```
Comment (killme2008) [inline src/servers/src/http/handler.rs:143]: What about move it into 

https://github.com/GreptimeTeam/greptimedb/blob/36b1bafbf00152cc89c8d4362ace432bbb1932a3/src/operator/src/statement.rs#L111

?

Then you can intercept all the queries.

**Review by killme2008 — COMMENTED** [review]
Good job! Another suggestion is that maybe we can record all the slow queries into a table like `information_schema.slow_queries` like that. Which is a append-only tables with TTL setting.

**Comment by zyy17** [comment]
> Good job! Another suggestion is that maybe we can record all the slow queries into a table like `information_schema.slow_queries` like that. Which is a append-only tables with TTL setting.



> Good job! Another suggestion is that maybe we can record all the slow queries into a table like `information_schema.slow_queries` like that. Which is a append-only tables with TTL setting.

I can take it as the next PR. Let me learn how to implement it.

File: config/config.md (hunk 1 above)
Comment (zyy17) [reply config/config.md:167]: Maybe it's not a necessary configuration? The current implementation is very cheap: log the slow query into file and collect by 3rd tools.

File: src/servers/src/http/handler.rs (hunk 3 above)
Comment (zyy17) [reply src/servers/src/http/handler.rs:148]: Seems a good idea. Let me have a try.

File: config/config.md (hunk 1 above)
Comment (killme2008) [reply config/config.md:167]: The sample ratio determines whether to print all slow queries, defaulting to 1 (100%). If all queries are slow, printing them all could harm database performance.

**Review by killme2008 — APPROVED** [review]
LGTM

**Comment by v0y4g3r** [comment]
Is it possible that we may print sensitive data into slow log file? Like `copy from/to` statements usually take a long time, while bearing AK/SK info.

**Review by fengjiachun — APPROVED** [review]
LGTM

**Comment by zyy17** [comment]
> Is it possible that we may print sensitive data into slow log file? Like `copy from/to` statements usually take a long time, while bearing AK/SK info.

You're right. Let me check it.

**Comment by zyy17** [comment]
> Is it possible that we may print sensitive data into slow log file? Like `copy from/to` statements usually take a long time, while bearing AK/SK info.

We already use https://github.com/GreptimeTeam/greptimedb/blob/main/src/sql/src/statements/option_map.rs#L27 for sensitive data. If we use `Display`, it will print `****` for senstive data.

**Pull request merged**